### PR TITLE
Feature/#23 implement budget account entity creation

### LINKED ISSUE #23: Implement `BudgetAccount` entity creation
In order to track current financial accounts - we need to provide user with a way to create their `Budget Account` - the entity that will represent user's income / debt / cash on hand / bank account / whatever else

Let's create Prisma model `BudgetAccount` with following fields:
- `id` - autogenerated UUID
- `userId` - string, required, corresponds to `User` entity
- `title` - string, required
- `description` - string, optional
- `icon` - string, optional
- `monthlyLimit` - Decimal, required, defaults to 0
- `type` - enum `BudgetAccountType` with following values `Income`, `Debt`, `SpendingCategory`, `Credit`, `Debit`. Defaults to `Debit`
- `user` - mapping to the `User` entity based on `userId` column

Then, after creating model - we should add button "Create Account" into `Budget` page. 
Clicking this button should open up `Dialog` with fields for `title`, `description`, `icon`, `monthlyLimit` and `type`.

Submitting form should show informational toast saying `Hand tight - we are creating budget account for ya...`.
On success - it should replace informational toast with success toast and close the modal
On error - it should replace informational toast with error toast and keep the modal opened

### PR DESCRIPTION
**Pull request by Artemh59**
Closes #23

### CODE REVIEW

File: prisma/schema.prisma [hunk 1]
```diff
@@ -59,3 +60,22 @@ model VerificationRequest {
 
   @@unique([identifier, token])
 }
+
+model BudgetAccount {
+  id           String            @id @default(uuid())
```
Comment (mudrila) [inline prisma/schema.prisma:65]: Let's use `cuid` to keep ID format consistent with `User` model - apologies for the confusion in the ticket description

File: prisma/schema.prisma [hunk 2]
```diff
@@ -59,3 +60,22 @@ model VerificationRequest {
...
+  userId       String
+  title        String
+  description  String?
+  icon         String?
+  monthlyLimit Decimal           @default(0)
```
Comment (mudrila) [inline prisma/schema.prisma:70]: Let's also add `currentBalance` field 
`currentBalance Decimal @default(0)`

File: prisma/schema.prisma [hunk 3]
```diff
@@ -59,3 +60,22 @@ model VerificationRequest {
...
+  type         BudgetAccountType @default(Debit)
+  user         User              @relation(fields: [userId], references: [id])
+}
+
+enum BudgetAccountType {
```
Comment (mudrila) [inline prisma/schema.prisma:75]: Let's also add `Goal` entry into `BudgetAccountType` - this will represent certain financial goal and will work slightly differently comparing to other types

File: src/app/api/budgetAccount/create/route.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+          title,
+          description,
+          icon,
+          monthlyLimit,
+          type,
+        },
+      });
+    }
+
+    return NextResponse.json({ message: 'Your account created!' });
```
Comment (mudrila) [inline src/app/api/budgetAccount/create/route.ts:46]: Let's just return `success: true` instead of message - it will be responsibility of FE side to decide which message to display in case of successful creation - there should be just clear indication that everything went well

File: src/app/api/budgetAccount/create/route.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  } catch (e) {
+    console.error(e, 'Error during account creation');
+    return NextResponse.json({
+      errorMessage: 'Please check our data',
```
Comment (mudrila) [inline src/app/api/budgetAccount/create/route.ts:50]: Let's rephrase it to `errorMessage: 'Unexpected error happened while creating budget account.'`

File: src/app/components/molecules/BudgetAccount/BudgetAccount.tsx [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+  const handleClose = () => {
+    setOpen(false);
+  };
+  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    toast.message('Hand tight - we are creating budget account for ya...');
+
+    const formData = new FormData(event.currentTarget);
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccount.tsx:35]: Let's use react controlled state / controlled inputs instead of native event DOM API

File: src/app/components/molecules/BudgetAccount/BudgetAccount.tsx [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    setOpen(true);
+  };
+
+  const handleClose = () => {
+    setOpen(false);
+  };
+  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    toast.message('Hand tight - we are creating budget account for ya...');
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccount.tsx:33]: I'd suggest `toast.info` - also we need to dismiss this toast once we've got response from the API route.

```
const loadingToastId = toast.info('...');

// blah blah making request

const responseData = await response.json();
toast.dismiss(loadingToastId);

if (responseData.success) {
  toast.success('Hell fucking yeah dude, it works!', { duration: 5000 });
} else if (responseData.errorMessage) {
  toast.error(responseData.errorMessage)
} else {
  toast.error('Something went wrong...');
}

```

File: src/app/components/molecules/BudgetAccount/BudgetAccount.tsx (hunk 6 above)
Comment (mudrila) [reply src/app/components/molecules/BudgetAccount/BudgetAccount.tsx:35]: Also please add `useTransition` usage to visualize loading state for the user

File: src/constants/content.ts [hunk 8]
```diff
@@ -7,5 +7,6 @@ export const PASSWORD_STRENGTH_DESCRIPTIONS = [
   'Very Strong',
 ];
 export const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
-export const passwordRegex =
-  /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&#])[A-Za-z\d@$!%*?&]{8,}$/;
+export const passwordRegex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[^A-Za-z\d])[A-Za-z\d\S]{8,}$/;
+
+export const accountTypes = ['Income', 'Debt', 'SpendingCategory', 'Credit', 'Debit'] as const;
```
Comment (mudrila) [inline src/constants/content.ts:12]: You can form this array via `Object.values(BudgetAccountType) as const`

File: src/app/components/molecules/BudgetAccount/BudgetAccount.tsx [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+            name="monthlyLimit"
+            margin="dense"
+            label="Monthly Limit"
+            type="number"
+            fullWidth
+            variant="outlined"
+            defaultValue={0}
+            required
+          />
+          <TextField
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccount.tsx:122]: Let's explicitly use `Select` component https://mui.com/material-ui/react-select/

File: src/app/components/molecules/BudgetAccount/BudgetAccount.tsx [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+      </Button>
+      <Dialog
+        open={open}
+        onClose={handleClose}
+        PaperProps={{
+          component: 'form',
+          onSubmit: handleSubmit,
+        }}
+      >
+        <DialogTitle>Create Account</DialogTitle>
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccount.tsx:79]: Let's be explicit - `Create Budget Account`

File: src/app/components/molecules/BudgetAccount/BudgetAccount.tsx [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        <DialogContent>
+          <DialogContentText>To create an account, please enter your data here.</DialogContentText>
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccount.tsx:81]: Don't think we need this description here - title should be sufficient, let's save some space :)

File: src/app/components/molecules/BudgetAccount/BudgetAccount.tsx
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccount.tsx:1]: Let's rename this file to `BudgetAccountForm`, but keep it under `BudgetAccount` folder

File: src/app/app/page.tsx [hunk 12]
```diff
@@ -1,10 +1,10 @@
...
 import { Metadata } from 'next';
+import BudgetAccountForm from '@/app/components/molecules/BudgetAccount/BudgetAccount';
 
 export const metadata: Metadata = {
   title: 'Dashboard | Fin-Plan',
 };
 
 export default function DashboardPage() {
-  return <Typography>I am Dashboard page placeholder</Typography>;
+  return <BudgetAccountForm />;
```
Comment (mudrila) [inline src/app/app/page.tsx:9]: Instead - please revert this change and move `BudgetAccountForm` usage to the `BudgetPage`

File: src/app/components/molecules/BudgetAccount/BudgetAccount.tsx [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+                value={option}
+              >
+                {option}
+              </MenuItem>
+            ))}
+          </TextField>
+        </DialogContent>
+        <DialogActions>
+          <Button onClick={handleClose}>Cancel</Button>
+          <Button type="submit">Create</Button>
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccount.tsx:145]: Let's make it `variant="contained"`

File: src/app/components/molecules/BudgetAccount/BudgetAccount.tsx [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+                key={option}
+                value={option}
+              >
+                {option}
+              </MenuItem>
+            ))}
+          </TextField>
+        </DialogContent>
+        <DialogActions>
+          <Button onClick={handleClose}>Cancel</Button>
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccount.tsx:144]: Let's make it `variant="outlined"` and also `type="reset"` to clear up the form

File: src/app/components/molecules/BudgetAccount/BudgetAccount.tsx [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    }
+  };
+
+  return (
+    <>
+      <Button
+        variant="outlined"
+        onClick={handleClickOpen}
+      >
+        Create Account
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccount.tsx:69]: Let's be explicit here - `Create Budget Account`
And also let's add Plus icon after the text through `endIcon` prop

Comment (mudrila) [reply src/app/components/molecules/BudgetAccount/BudgetAccount.tsx:69]: https://mui.com/material-ui/react-button/#buttons-with-icons-and-label

File: src/app/components/molecules/BudgetAccount/BudgetAccount.tsx [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+          <TextField
+            id="description"
+            name="description"
+            margin="dense"
+            label="Description (Optional)"
+            type="text"
+            fullWidth
+            variant="outlined"
+          />
+          <TextField
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccount.tsx:102]: For this one - let's replicate following UI/UX
https://mui.com/material-ui/material-icons/#search-material-icons
The only thing is that it should show nested modal with icon selection.
The value supplied should be an unique ID of the icon, from which we could easily identify which icon should be shown in the UI

File: src/app/components/molecules/BudgetAccount/BudgetAccount.tsx [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+
+  return (
+    <>
+      <Button
+        variant="outlined"
+        onClick={handleClickOpen}
+      >
+        Create Account
+      </Button>
+      <Dialog
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccount.tsx:71]: Please, make sure this dialog takes full screen width below `md` resolution

Comment (mudrila) [reply src/app/components/molecules/BudgetAccount/BudgetAccount.tsx:71]: At the same time - above `md` it should be statically `600px` width

**Review by mudrila — CHANGES_REQUESTED** [review]
Quite a bit of stuff to do sir - but pretty good start sir!

File: src/constants/content.ts (hunk 8 above)
Comment (Artemh59) [reply src/constants/content.ts:12]: idk I can't do it properly, it don't work with enum in route.ts
can I just add 'Goal'?

File: src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+            fullWidth
+            maxWidth="md"
+          >
+            <DialogTitle>Select an Icon</DialogTitle>
+            <DialogContent>
+              <Grid2
+                container
+                spacing={2}
+              >
+                {Object.keys(Icons).map(iconName => (
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx:168]: Oh yeah - we cannot render the whole damn thing at once - we need virtualization to prevent application lagging
Please add something like this: https://github.com/bvaughn/react-window

File: src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+            margin="dense"
+            label="Description (Optional)"
+            type="text"
+            fullWidth
+            variant="outlined"
+            value={description}
+            onChange={e => setDescription(e.target.value)}
+            disabled={isPending}
+          />
+          <TextField
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx:138]: Please, extract this icon selection into `IconSelect` component

File: src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx [hunk 20]
```diff
@@ -0,0 +1,246 @@
...
+    startTransition(async () => {
+      const response = await fetch('/api/budgetAccount/create', {
+        method: 'POST',
+        body: JSON.stringify({
+          title,
+          description,
+          icon,
+          monthlyLimit,
+          type,
+          userId,
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx:70]: We don't need to pass `userId` from here - we can read it within API route via `const session = await auth()` usage

Comment (mudrila) [reply src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx:70]: And that will be safer - otherwise anyone could just create budget account via specifying someone's user ID

**Review by mudrila — CHANGES_REQUESTED** [review]
Almost there sir!
Apologies - haven't mentioned we would need virtualization to improve app performance while rendering large list of icons.
Also, we shouldn't be passing `userId` from FE to BE - it should be retrieved through `auth()` usage

File: src/app/components/molecules/BudgetAccount/BudgetAccount.tsx (hunk 16 above)
Comment (mudrila) [reply src/app/components/molecules/BudgetAccount/BudgetAccount.tsx]: We'll need to use virtualization in order to make it performant
I'd albeit from hand-picking icons at this stage - yes, it is reasonable in longer run - but I don't want us be spending time and picking/adjusting available icons each time we need it - just allow any icon and we'll introduce more tailored icon selection later on :)

File: src/constants/content.ts (hunk 8 above)
Comment (mudrila) [reply src/constants/content.ts]: Yeah, no worries - we'll figure this out later - minor nitpick

File: src/app/api/budgetAccount/create/route.ts [hunk 21]
```diff
@@ -0,0 +1,57 @@
+import { NextResponse } from 'next/server';
+import { z } from 'zod';
+import { accountTypes } from '@/constants/content';
+import prisma from '@/utils/prisma';
+
+export async function POST(request: Request) {
+  try {
+    const { title, description, icon, monthlyLimit, type, currentBalance, userId } =
```
Comment (mudrila) [inline src/app/api/budgetAccount/create/route.ts:8]: So, instead of extracting `userId` from here - please add `const session = await auth()` usage and validate that session / user actually present - return error message in the response if that's not the case

File: src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx [hunk 22]
```diff
@@ -0,0 +1,261 @@
...
+    return IconComponent ? <IconComponent /> : null;
+  };
+
+  const IconRow = ({ index, style }: { index: number; style: React.CSSProperties }) => {
+    const startIndex = index * iconsPerRow;
+    const endIndex = Math.min(startIndex + iconsPerRow, iconList.length);
+    const iconsForRow = iconList.slice(startIndex, endIndex);
+
+    return (
+      <div style={{ ...style, display: 'flex', justifyContent: 'space-around' }}>
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx:97]: No explicit inline styling sir - please, transform this into `Box` element and pass `style={style}`
Unless it doesn't work that way - if so - let me know

File: src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx [hunk 23]
```diff
@@ -0,0 +1,261 @@
...
+      }
+    });
+  };
+
+  const renderIcon = (iconName: string) => {
+    const IconComponent = Icons[iconName as keyof typeof Icons];
+    return IconComponent ? <IconComponent /> : null;
+  };
+
+  const IconRow = ({ index, style }: { index: number; style: React.CSSProperties }) => {
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx:91]: Please extract it into separate file `src/components/molecules/IconSelect/IconSelect.tsx` and import from there

File: src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx [hunk 24]
```diff
@@ -0,0 +1,261 @@
...
+      } else if (responseData.success) {
+        toast.success('Your budget account created!', { duration: 5000 });
+        handleClose();
+      } else {
+        toast.error('Something went wrong...');
+      }
+    });
+  };
+
+  const renderIcon = (iconName: string) => {
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx:86]: Same with this function - put it there please and export

Comment (mudrila) [reply src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx:86]: Also would be better to transform it into component

```
function IconRenderrer({ iconName }: { iconName: string}) {
  const IconComponent = Icons[iconName as keyof typeof Icons];
  return IconComponent ? <IconComponent /> : null;
}
```

Will make react little bit more performant on working with DOM

File: src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx (hunk 23 above)
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx:91]: `React.CSSProperties` -> `CSSProperties` and import that from `react`

File: src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx [hunk 25]
```diff
@@ -0,0 +1,261 @@
...
+            margin="dense"
+            label="Description (Optional)"
+            type="text"
+            fullWidth
+            variant="outlined"
+            value={description}
+            onChange={e => setDescription(e.target.value)}
+            disabled={isPending}
+          />
+          <FormControl
```
Comment (mudrila) [inline src/app/components/molecules/BudgetAccount/BudgetAccountForm.tsx:156]: Again - please, extract this `FormControl` into separate component under `src/components/molecules/IconSelect/IconSelect.tsx`

We're certainly will be re-using this component, but also this will make budget account form cleaner

**Review by mudrila — CHANGES_REQUESTED** [review]
Couple final nitpicks sir

**Review by mudrila — APPROVED** [review]
